polkadot/xcm/pallet-xcm/src/lib.rs

```diff
@@ -1648,33 +1692,101 @@ impl<T: Config> Pallet<T> {
 		}
 		let fees = assets.get(fee_asset_item as usize).ok_or(Error::<T>::Empty)?.clone();
 
-		Self::build_and_execute_xcm_transfer_type(
-			origin_location,
-			dest,
+		let (local_xcm, remote_xcm) = Self::build_xcm_transfer_type(
+			origin_location.clone(),
+			dest.clone(),
 			beneficiary,
 			assets,
 			TransferType::Teleport,
 			FeesHandling::Batched { fees },
 			weight_limit,
-		)
+		)?;
+		Self::execute_xcm_transfer(origin_location, dest, local_xcm, remote_xcm)
+	}
+
+	fn do_transfer_assets(
+		origin: Location,
+		dest: Location,
+		beneficiary: Location,
+		mut assets: Vec<Asset>,
+		assets_transfer_type: TransferType,
+		fees: Asset,
+		fees_transfer_type: TransferType,
+		weight_limit: WeightLimit,
+	) -> DispatchResult {
+		// local and remote XCM programs to potentially handle fees separately
+		let fees = if fees_transfer_type == assets_transfer_type {
+			// no need for custom fees instructions, fees are batched with assets
+			FeesHandling::Batched { fees }
+		} else {
+			// Disallow _remote reserves_ unless assets & fees have same remote reserve (covered
+			// by branch above). The reason for this is that we'd need to send XCMs to separate
+			// chains with no guarantee of delivery order on final destination; therefore we
+			// cannot guarantee to have fees in place on final destination chain to pay for
+			// assets transfer.
+			ensure!(
+				!matches!(assets_transfer_type, TransferType::RemoteReserve(_)),
+				Error::<T>::InvalidAssetUnsupportedReserve
+			);
+			let weight_limit = weight_limit.clone();
+			// remove `fees` from `assets` and build separate fees transfer instructions to be
+			// added to assets transfers XCM programs
+			let fee_asset_index =
+				assets.iter().position(|a| a.id == fees.id).ok_or(Error::<T>::FeesNotMet)?;
+			let fees = assets.remove(fee_asset_index);
```

Comment:
I am not sure here, if this is a problem, but input `fees: Asset` amount is ignored in this branch, just the `assets.remove(fees.id);`, 

I am just thinking if we shouldn't just first remove `fees: Asset` amount from `assets` and when `0`, then remove  the whole asset by index.

Something like this pseudo-code:
```
let fees = {
    // find `fees.id` in `assets` -> found_fees_asset
    let mut found_fees_asset = assets.get_mut(fees.id)
   
    // just remove amount of `fees` from `found_fees_asset`
    found_fees_asset.fun = found_fees_asset.fun - fees.fun;

    // if we removed all amount, then also remove `fees.id` from `assets`
    if found_fees_asset.fun == 0 {
        assets.remove(fees.id);
    }

    // return untouched `input `fees: Asset``
    fees
} 
```

E.g. `AssetsInHolding` has similar function for removing Asset [`fn checked_sub(mut self, asset: Asset)`](https://github.com/paritytech/polkadot-sdk/blob/master/polkadot/xcm/xcm-executor/src/assets.rs#L414)

---
For example, iiuc, I want to transfer this:
```
transfer_assets_using_type(
assets: [assets1(1000 tokens), assets2(5000 tokens)]
fees: [assets1(200 tokens)]
)
```

Just removing the index results in transferring:
```
assets: [assets2(5000 tokens)]
fees: [assets1(200 tokens)]
```

With my suggestion, it would result in transferring:
```
assets: [assets1(800 tokens), assets2(5000 tokens)]
fees: [assets1(200 tokens)]
```

Reply:
right, good catch!

Reply:
actually did your other suggestion: https://github.com/paritytech/polkadot-sdk/pull/3695#discussion_r1561256820